File: 0421-Find-Maximum-XOR/0421.py
class Solution:
  def findMaximumXOR(self, nums: List[int]) -> int:
    HIGH_BIT = 30

    x = 0
    for k in range(HIGH_BIT, -1, -1):
      seen = set()
      for num in nums:
        seen.add(num >> k)

      x_next = x * 2 + 1
      found = False

      for num in nums:
        if x_next ^ (num >> k) in seen:
          found = True
          break

      if found:
        x = x_next
      else:
        x = x_next - 1

    return x

File: 0421-Find-Maximum-XOR/0421.go
func findMaximumXOR(nums []int) (x int) {
	const highBit = 30
	for k := highBit; k >= 0; k-- {
		seen := map[int]bool{}
		for _, num := range nums {
			seen[num>>k] = true
		}

		xNext = x*2 + 1
		found := false

		for _, num := range nums {
			if seen[num>>k^xNext] {
				found = true
				break
			}
		}

		if found {
			x = xNext
		} else {
			x = xNext - 1
		}
	}
	return
}